expectations.go

```diff
@@ -134,20 +134,48 @@ type queryBasedExpectation struct {
 	args      []interface{}
 }
 
-func (e *queryBasedExpectation) argsMatches(args []interface{}) error {
-	if len(args) != len(e.args) {
-		return fmt.Errorf("expected %d, but got %d arguments", len(e.args), len(args))
+func (e *queryBasedExpectation) argsMatches(sql string, args []interface{}) error {
+	eargs := e.args
+	// check for any query rewriters
+	// according to current pgx docs, a QueryRewriter is only supported as the first
+	// argument.
+	if len(args) == 1 {
+		if qrw, ok := args[0].(pgx.QueryRewriter); ok {
+			// note: pgx.Conn is not currently used by the query rewriter, but is part
+			// of the method signature, so just create an empty pointer for now.
+			_, newArgs, err := qrw.RewriteQuery(context.Background(), new(pgx.Conn), sql, args)
```

Comment:
I'm thinking if we should save a `newSQL` value here, so latter we can do things like this:


```go
  	mock.ExpectExec("INSERT INTO users(name, created_at) VALUES (@name, @time)").
		WithArgs(pgx.NamedArgs{"name": "john", "time": AnyTime{}}).
                WithRewrittenQuery("INSERT INTO users(name, created_at) VALUES ($1, $2)").
		WillReturnResult(NewResult("INSERT", 1)).

	_, err = mock.Exec(context.Background(),
		"INSERT INTO users(name, created_at) VALUES (@name, @time)",
		pgx.NamedArgs{"name": "john", "time": time.Now()},
	)
```

So the idea is to provide a possibility to control the actual query sent to the server. What do you think?

Reply:
When you say "control the actual query sent to the server", what do you mean (the mock never sends to a server, does it)?
Would it be used an expectation for ensuring the named arguments get substituted for `$` values correctly?

Reply:
Yeah, it should be quoted, "sent". The query rewriter is very powerful. I saw an example with empty initial query made on fly by the struct (https://ectobit.com/blog/pgx-v5-2/).

So I started to think that developers would like to have the possibility to control both: initial and rewritten query in tests.

Reply:
This would be the proper example

```go
  	mock.ExpectExec("").
		WithArgs(pgx.NamedArgs{"name": "john", "time": AnyTime{}}).
                WithRewrittenQuery("INSERT INTO users(name, created_at) VALUES ($1, $2)").
		WillReturnResult(NewResult("INSERT", 1)).

	_, err = mock.Exec(context.Background(),
		"INSERT INTO users(name, created_at) VALUES (@name, @time)",
		pgx.NamedArgs{"name": "john", "time": time.Now()},
	)

Reply:
Hmm. I wonder if `WithRewrittenQuery` could accept a function like..
`func (rewrittenSQL string) error` where if the return error is not nil, the check fails. This could allow someone to validate the sql as they desire (for example with `pg_query_go` to ensure correct parsing)?

Reply:
> This would be the proper example
> 
> ```go
>   	mock.ExpectExec("").
> 		WithArgs(pgx.NamedArgs{"name": "john", "time": AnyTime{}}).
>                 WithRewrittenQuery("INSERT INTO users(name, created_at) VALUES ($1, $2)").
> 		WillReturnResult(NewResult("INSERT", 1)).
> 
> 	_, err = mock.Exec(context.Background(),
> 		"INSERT INTO users(name, created_at) VALUES (@name, @time)",
> 		someUserStruct{"name": "john", "time": time.Now()},
> 	)
> ```

aha. I see what you mean now. I was thinking WithRewrittenQuery as an expectation, and you were thinking of it as the sql to match against (instead of using the sql in ExpectExec).

Reply:
Yes, it's an expectation. I expect initial query and I might want to expect rewritten query. Just like with arguments, one can use them or just write plain query